csrc/python_frontend/fusion_record.h

```diff
@@ -611,6 +611,93 @@ struct PermuteOpRecord : RecordFunctor {
   std::vector<int64_t> dims_;
 };
 
+// TODO: should this one be merged with PermuteOpRecord?
+struct SetStrideOrderOpRecord : RecordFunctor {
+  SetStrideOrderOpRecord(
+      std::vector<State> _args,
+      std::vector<State> _outputs,
+      std::vector<int64_t> stride_order)
+      : RecordFunctor(
+            std::move(_args),
+            std::move(_outputs),
+            "ops.set",
+            serde::RecordType_SetStrideOrderOp),
+        stride_order_(std::move(stride_order)) {}
+  ~SetStrideOrderOpRecord() override = default;
+  RecordFunctor* clone() final {
+    return new SetStrideOrderOpRecord(*this);
+  }
+
+  size_t hash() const final {
+    auto result = RecordFunctor::hash();
+    size_t dims_hash = 0;
+    for (auto dim : stride_order_) {
+      hashCombine(dims_hash, static_cast<size_t>(dim));
+    }
+    return result | (dims_hash & 0xffff);
+  }
+
+  bool operator==(const RecordFunctor& other) const final {
+    auto result = false;
+    if (auto child_ptr = dynamic_cast<const SetStrideOrderOpRecord*>(&other)) {
+      result = RecordFunctor::operator==(other);
+      if (result) {
+        result = (stride_order_.size() == child_ptr->stride_order_.size());
+        if (result) {
+          for (size_t i = 0; i < stride_order_.size(); ++i) {
+            if (stride_order_[i] != child_ptr->stride_order_[i]) {
+              result = false;
+              break;
+            }
+          }
+        }
+      }
+    }
+    return result;
+  }
+
+  void operator()(FusionState& fd) final {
+    auto arg = fd.getFusionState(args_.at(0).index)->template as<TensorView>();
+    auto output = set(arg);
+    int64_t rank = static_cast<int>(stride_order_.size());
+    std::vector<IterDomain*> allocation_domain(rank);
+    for (auto i : c10::irange(rank)) {
+      allocation_domain[rank - 1 - stride_order_[i]] = output->axis(i);
+    }
+    output->setAllocationDomain(allocation_domain, true);
```

Comment:
IIUC, setting output stride is as simple as setting allocation domain on the tensor.

Note that user could also use this to manipulate the allocation domain on intermediate tensor, would that be causing any issue? I'm imagining this would just be ignored by codegen. cc'ing @zasdfgbnm

Reply:
If we want to restrict it to outputs only, could we add stride order as an optional arg to `add_output` instead?

Reply:
We already have the stride order as part of the API for add_output.

It's a bit tricky here how `stride_order` can be specified in framework now. i.e. framework might want to have an operator like `contiguous(stride_order)` to dictate the tensor stride. If we decide to support that op, it could ended up being in an arbitrary position in the fusion.

Reply:
Should we just ignore `contiguous` and only look at stride order of output? I don't understand what this is trying to support.

Reply:
This is just to keep our life simpler during parsing framework graphs.

i.e. framework could specify `x.contigous(channels_last)` in arbitrary places, not necessarily on a fusion output. Even though we would only care about contiguous calls on output and safely ignore any other, having a `set` op where we can specify output strides makes it easier so we don't need extra bits for bookkeeping.

Reply:
Then does it make sense to always make `contiguous` a set op?

Reply:
No, we do want to specify the right stride order.

i.e. we could have something like
```
t1 = normalization(t0)
t2 = contiguous(t1, channels_last)
convolution_out = convolution(t2)
```

So partition would give a nvfuser region like
```
t1 = normalization(t0)
t2 = contiguous(t1, channels_last)
# fusion.add_output(t2)
```

if we have a `set` op with stride order argument, when we parse the graph, we can simply just parse each op as-is, without tracing whether the graph is trying to set a stride-order on output or just intermediate tensor.